[Merged by Bors] - feat(RingTheory): add HahnSeries.truncLT

wwylele opened the pull request:
---

This is for the proof of Hahn embedding theorem #27043, but it feels niche to have it in HahnSeries file, so I am not sure... if this doesn't feel right, I could make them private to the proof.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

wwylele committed:
feat(RingTheory): add HahnSeries.truncate

ScottCarnahan commented on Mathlib/RingTheory/HahnSeries/Addition.lean:
```diff
@@ -527,6 +527,17 @@ def embDomainLinearMap (f : Γ ↪o Γ') : HahnSeries Γ R →ₗ[R] HahnSeries
+    ext i
+    by_cases h : c ≤ i <;> simp [h]
+  map_smul' s x := by
+    ext i
```
You don't really need `i` here.
```suggestion
    ext
```

ScottCarnahan reviewed (approved):
Looks good to me.

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Addition.lean

Co-authored-by: Scott Carnahan <128885296+ScottCarnahan@users.noreply.github.com>

Ruben-VandeVelde reviewed (approved):
maintainer merge

kbuzzard commented on Mathlib/RingTheory/HahnSeries/Basic.lean:
```diff
@@ -524,4 +524,24 @@ theorem order_ofForallLtEqZero [Zero Γ] (f : Γ → R) (hf : f ≠ 0) (n : Γ)
+
+/-- Zeros out coefficients of a `HahnSeries` at indices equal to or greater than `c`. -/
+@[simps]
+def truncate [PartialOrder Γ] (c : Γ) (x : HahnSeries Γ R) : HahnSeries Γ R where
+  coeff i := open Classical in if c ≤ i then 0 else x.coeff i
+  isPWO_support' := Set.IsPWO.mono x.isPWO_support (by simp)
```
I am a bit nervous about `open Classical in` in a definition (this is not a proof). If Gamma is something like the integers then there is a constructive algorithm for `c ≤ i` and this approach will I think override it and use LEM which might lead to diamonds later? I think it might be better to assume `[DecidableLE Γ]` in the definition.

wwylele committed:
use DecidableLE

wwylele committed:
Multiple update

- rename to trunc (similar to PowerSeries.trunc)
- upgrade the base trunc to a ZeroHom, similar to HahnSeries.single
- also add the AddMonoidHom version
- manual simp lemma to get the correct name

wwylele committed:
fix space

YaelDillies reviewed (commented):
Apologies, I don't actually know the proof of the Hahn embedding theorem. Would you mind explaining how truncation of Hahn series comes in? Naïvely, I would have expected truncation to be upwards, but instead it's downwards.

wwylele commented:
@YaelDillies sure, let me see if I can explain this well...

The proof I am following is https://www.ams.org/journals/proc/1952-003-06/S0002-9939-1952-0052045-1/. (The proof shown there is for a special case where the ordered group is a real vector space. Generalization is explained [here](https://www.jstor.org/stable/2032549?origin=crossref)). The truncation operation is called "cut" on page 980.

The proof is a process of transfinite induction starting from a partial map and adding new elements to the domain (Theorem 3.2). Whenever a new element is added, one needs to assign appropriate output to that element and preserve desired properties (strict monotone & linear). This turns out pretty hard on its own (maybe impossible?), so one strengthen the induction hypothesis with an extra condition on the partial map $F$:
 - if $f \in Range(F)$, then $Trunc(f, \Gamma) \in Range (F)$ for all $\Gamma$

Because of linearity of $F$, this can be rephrased in terms of the upward truncation $Trunc'(f, \Gamma) = f - Trunc(f, \Gamma) \in Range (F) $. I doubt that it will make the proof simpler, though.

This condition means that all values of $F$ is a limit point of a sequence by progressively adding more terms. It goes hand-in-hand with how $F(x)$ for a new element $x$ is chosen: we progressively add terms to $F(x)$ equal to a neighbor $F(y)$ in a ball defined by Archimedean class. if there is no such neighbor $y$ with $F(y)$ defined,  we stop adding more terms to $F(x)$ and leave the rest as zero. One can then show that $F(x)$ respects the order embedding: as $x$ slightly deviates from $y$, $F(x)$ does so from $F(y)$ in the same direction.

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Addition.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Addition.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Addition.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Basic.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

wwylele committed:
remove addMonoidHom

wwylele commented on Mathlib/RingTheory/HahnSeries/Basic.lean:
```diff
@@ -537,4 +537,33 @@ theorem order_ofForallLtEqZero [Zero Γ] (f : Γ → R) (hf : f ≠ 0) (n : Γ)
+/-- Zeroes out coefficients of a `HahnSeries` at indices equal to or greater than `c`. -/
+def trunc [PartialOrder Γ] [DecidableLE Γ] (c : Γ) : ZeroHom (HahnSeries Γ R) (HahnSeries Γ R) where
+  toFun x :=
+    { coeff i := if c ≤ i then 0 else x.coeff i
```
Assuming the LE is not a typo, then matters, The \Gamma type can be as large as Real, where closed interval and open interval are different.
Writing it as `i < c` instead of `i <= c` would be fine for me, though I get a feeling that mathlib generally prefers LE over LT?

wwylele replied:
... Thinking about it again, it probably doesn't matter for the proof I am working on, but what's written here is the convention used in the proof, and it also matches Polynomial.trunc

wwylele replied:
I will go with 2 then, as I think I can come up with pathological object that will break 1 soon
Perhaps I should also change this to LinearOrder, which is what I really care about

wwylele committed:
addressed comments

wwylele committed:
update doc

YaelDillies commented on Mathlib/RingTheory/HahnSeries/Basic.lean:
```diff
@@ -537,4 +537,31 @@ theorem order_ofForallLtEqZero [Zero Γ] (f : Γ → R) (hf : f ≠ 0) (n : Γ)
+    (c : Γ) (x : HahnSeries Γ R) (i : Γ) :
+    (truncLT c x).coeff i = if i < c then x.coeff i else 0  := rfl
+
+theorem coeff_truncLT_eq [PartialOrder Γ] [DecidableLT Γ]
```
```suggestion
theorem coeff_truncLT_of_lt [PartialOrder Γ] [DecidableLT Γ]
```

YaelDillies commented on Mathlib/RingTheory/HahnSeries/Basic.lean:
```diff
@@ -537,4 +537,31 @@ theorem order_ofForallLtEqZero [Zero Γ] (f : Γ → R) (hf : f ≠ 0) (n : Γ)
+    {c i : Γ} (h : i < c) (x : HahnSeries Γ R) : (truncLT c x).coeff i = x.coeff i := by
+  simp [h]
+
+theorem coeff_truncLT_eq_zero [LinearOrder Γ]
```
```suggestion
theorem coeff_truncLT_of_le [LinearOrder Γ]
```

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Basic.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

wwylele committed:
Update Mathlib/RingTheory/HahnSeries/Basic.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

riccardobrasca commented on Mathlib/RingTheory/HahnSeries/Addition.lean:
```diff
@@ -549,6 +549,20 @@ def embDomainLinearMap (f : Γ ↪o Γ') : HahnSeries Γ R →ₗ[R] HahnSeries
+/-- `HahnSeries.truncLT` as a linear map. -/
+def truncLTLinearMap [DecidableLT Γ] (c : Γ) : HahnSeries Γ V →ₗ[R] HahnSeries Γ V where
+  toFun := truncLT c
+  map_add' x y := by
```
Can you prove these two results as separated lemmas (with minimal assumptions)? Thanks!

riccardobrasca reviewed (approved):
Thanks!

bors d+

alreadydone commented on Mathlib/RingTheory/HahnSeries/Basic.lean:
```diff
@@ -537,4 +537,31 @@ theorem order_ofForallLtEqZero [Zero Γ] (f : Γ → R) (hf : f ≠ 0) (n : Γ)
+section Truncate
+variable [Zero R]
+
+/-- Zeroes out coefficients of a `HahnSeries` at indices not less than `c`. -/
```
It seems the singular form of the verb "zero" is "zeros": https://en.wiktionary.org/wiki/zeros_in

wwylele committed:
extract theorems

wwylele committed:
Merge remote-tracking branch 'upstream/master' into wwylele-hahn-truncate

wwylele commented:
bors r+
